components/g3d/include/g3d_file.h

```diff
@@ -0,0 +1,192 @@
+/* openblack - A reimplementation of Lionhead's Black & White.
+ *
+ * openblack is the legal property of its developers, whose names
+ * can be found in the AUTHORS.md file distributed with this source
+ * distribution.
+ *
+ * openblack is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * openblack is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with openblack. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <istream>
+#include <map>
+#include <streambuf>
+#include <string>
+#include <vector>
+
+namespace openblack::g3d
+{
+
+// From https://stackoverflow.com/a/13059195/10604387
+struct membuf: std::streambuf
+{
+	membuf(char const* base, size_t size)
+	{
+		char* p(const_cast<char*>(base));
+		this->setg(p, p, p + size);
+	}
+};
+
+struct imemstream: virtual membuf, std::istream
+{
+	imemstream(char const* base, size_t size)
+		: membuf(base, size)
+		, std::istream(static_cast<std::streambuf*>(this))
+	{
+	}
+};
```

Comment:
This is only used internally, move it to cpp file, sync with l3d's version which is more advanced